# APIs Lab
In this lab we will practice using APIs to retrieve and store data.

In [1]:
# Imports at the top
import json
import urllib
import pandas as pd
import numpy as np
import requests
import json
import re
import matplotlib.pyplot as plt
%matplotlib inline

## Exercise 1: Get Data From Sheetsu

[Sheetsu](https://sheetsu.com/) is an online service that allows you to access any Google spreadsheet from an API. This can be a very handy way to share a dataset with colleagues as well as to create a mini centralized data storage, that is simpler to edit than a database.

A Google Spreadsheet with wine data can be found [here](https://docs.google.com/a/generalassemb.ly/spreadsheets/d/1JWRwDnwIMLgvPqNMdJLmAJgzvz0K3zAUc6jev3ci1c8/edit?usp=sharing).

You can access it through the Sheetsu API at this endpoint: https://sheetsu.com/apis/v1.0/cc9420722ae4. [Here](https://sheetsu.com/docs/beta) is Sheetsu's documentation.


Questions:

1. Use the requests library to access the document. Inspect the response text. What kind of data is it?
2. Check the status code of the response object. What code is it?
3. Use the appropriate libraries and read functions to read the response into a Pandas Dataframe
4. Once you've imported the data into a dataframe, check the value of the 5th line: what's the price?

In [2]:
url = 'https://sheetsu.com/apis/v1.0/cc9420722ae4'
response = requests.get(url)

In [3]:
response.text

u'[{"Color":"W","Region":"Portugal","Country":"Portugal","Vintage":"2013","Vinyard":"Vinho Verde","Name":"","Grape":"","Consumed In":"2015","Score":"4","Price":""},{"Color":"W","Region":"France","Country":"France","Vintage":"2013","Vinyard":"Peyruchet","Name":"","Grape":"","Consumed In":"2015","Score":"3","Price":"17.8"},{"Color":"W","Region":"Oregon","Country":"Oregon","Vintage":"2013","Vinyard":"Abacela","Name":"","Grape":"","Consumed In":"2015","Score":"3","Price":"20"},{"Color":"W","Region":"Spain","Country":"Spain","Vintage":"2012","Vinyard":"Ochoa","Name":"","Grape":"chardonay","Consumed In":"2015","Score":"2.5","Price":"7"},{"Color":"R","Region":"","Country":"US","Vintage":"2012","Vinyard":"Heartland","Name":"Spice Trader","Grape":"chiraz, cab","Consumed In":"2015","Score":"3","Price":"6"},{"Color":"R","Region":"California","Country":"US","Vintage":"2012","Vinyard":"Crow Canyon","Name":"","Grape":"cab","Consumed In":"2015","Score":"3.5","Price":"13"},{"Color":"R","Region":"Orego

In [4]:
response

<Response [200]>

In [5]:
response.json()
type(response.json())

list

In [6]:
r = pd.read_json(response.text)

In [7]:
r.head(6)

Color  Consumed In   Country        Grape          Name Price      Region  \
0     W         2015  Portugal                                     Portugal   
1     W         2015    France                             17.8      France   
2     W         2015    Oregon                               20      Oregon   
3     W         2015     Spain    chardonay                   7       Spain   
4     R         2015        US  chiraz, cab  Spice Trader     6               
5     R         2015        US          cab                  13  California   

  Score  Vintage      Vinyard  
0     4     2013  Vinho Verde  
1     3     2013    Peyruchet  
2     3     2013      Abacela  
3   2.5     2012        Ochoa  
4     3     2012    Heartland  
5   3.5     2012  Crow Canyon

In [8]:
jr = json.loads(response.text)

In [9]:
jr = pd.DataFrame(jr)

In [10]:
r['Price'].iloc[5]

u'13'

> Answers:
    1. A JSON string.
    2. 200
    3. Options inlucde: pd.read_json; json.loads + pd.Dataframe
    4. 5

### Exercise 2: Post Data to Sheetsu
Now that we've learned how to read data, it'd be great if we could also write data. For this we will need to use a _POST_ request.

1. Use the post command to add the following data to the spreadsheet:

In [11]:
post_data = {
'Grape' : ''
, 'Name' : 'My wine'
, 'Color' : 'R'
, 'Country' : 'US'
, 'Region' : 'Sonoma'
, 'Vinyard' : ''
, 'Score' : '10'
, 'Consumed In' : '2015'
, 'Vintage' : '1973'
, 'Price' : '200'
}

In [12]:
requests.post(url, data = post_data)

<Response [201]>

1. What status did you get? How can you check that you actually added the data correctly?
- In this exercise, your classmates are adding data to the same spreadsheet. What happens because of this? Is it a problem? How could you mitigate it?

201 (data created) - check documentation that it added

## Exercise 3: Data munging

Get back to the dataframe you've created in the beginning. Let's do some data munging:

1. Search for missing data
    - Is there any missing data? How do you deal with it?
    - Is there any data you can just remove?
    - Are the data types appropriate?
- Summarize the data 
    - Try using describe, min, max, mean, var

In [14]:
r.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 67 entries, 0 to 66
Data columns (total 10 columns):
Color          67 non-null object
Consumed In    67 non-null int64
Country        67 non-null object
Grape          67 non-null object
Name           67 non-null object
Price          67 non-null object
Region         67 non-null object
Score          67 non-null object
Vintage        67 non-null int64
Vinyard        67 non-null object
dtypes: int64(2), object(8)
memory usage: 5.8+ KB


In [20]:
r_emp = r.replace('', np.nan)

In [22]:
r_emp.head()

Color  Consumed In   Country        Grape          Name Price    Region  \
0     W         2015  Portugal          NaN           NaN   NaN  Portugal   
1     W         2015    France          NaN           NaN  17.8    France   
2     W         2015    Oregon          NaN           NaN    20    Oregon   
3     W         2015     Spain    chardonay           NaN     7     Spain   
4     R         2015        US  chiraz, cab  Spice Trader     6       NaN   

  Score  Vintage      Vinyard  
0     4     2013  Vinho Verde  
1     3     2013    Peyruchet  
2     3     2013      Abacela  
3   2.5     2012        Ochoa  
4     3     2012    Heartland

In [29]:
r_emp = r_emp[pd.isnull(r_emp).any(axis=1)]
r_emp.head()

Color  Consumed In   Country        Grape          Name Price    Region  \
0     W         2015  Portugal          NaN           NaN   NaN  Portugal   
1     W         2015    France          NaN           NaN  17.8    France   
2     W         2015    Oregon          NaN           NaN    20    Oregon   
3     W         2015     Spain    chardonay           NaN     7     Spain   
4     R         2015        US  chiraz, cab  Spice Trader     6       NaN   

  Score  Vintage      Vinyard  
0     4     2013  Vinho Verde  
1     3     2013    Peyruchet  
2     3     2013      Abacela  
3   2.5     2012        Ochoa  
4     3     2012    Heartland

In [31]:
r_emp['Color'].unique()

array([u'W', u'R', u'P'], dtype=object)

## Exercise 4: Feature Extraction

We would like to use a regression tree to predict the score of a wine. In order to do that, we first need to select and engineer appropriate features.

- Set the target to be the Score column, drop the rows with no score
- Use pd.get_dummies to create dummy features for all the text columns
- Fill the nan values in the numerical columns, using an appropriate method
- Train a Decision tree regressor on the Score, using a train test split:
        X_train, X_test, y_train, y_test, = train_test_split(X, y, test_size=0.3, random_state=42)
- Plot the test values, the predicted values and the residuals
- Calculate R^2 score
- Discuss your findings


## Exercise 5: IMDB Movies

Sometimes an API doesn't provide all the information we would like to get and we need to be creative.
Here we will use a combination of scraping and API calls to investigate the ratings and gross earnings of famous movies.

## 5.a Get top movies

The Internet Movie Database contains data about movies. Unfortunately it does not have a public API.

The page http://www.imdb.com/chart/top contains the list of the top 250 movies of all times. Retrieve the page using the requests library and then parse the html to obtain a list of the `movie_ids` for these movies. You can parse it with regular expression or using a library like `BeautifulSoup`.

**Hint:** movie_ids look like this: `tt2582802`

## 5.b Get top movies data

Although the Internet Movie Database does not have a public API, an open API exists at http://www.omdbapi.com.

Use this API to retrieve information about each of the 250 movies you have extracted in the previous step.
- Check the documentation of omdbapi.com to learn how to request movie data by id
- Define a function that returns a python object with all the information for a given id
- Iterate on all the IDs and store the results in a list of such objects
- Create a Pandas Dataframe from the list

## 5.c Get gross data

The OMDB API is great, but it does not provide information about Gross Revenue of the movie. We'll revert back to scraping for this.

- Write a function that retrieves the gross revenue from the entry page at imdb.com
- The function should handle the exception of when the page doesn't report gross revenue
- Retrieve the gross revenue for each movie and store it in a separate dataframe

## 5.d Data munging

- Now that you have movie information and gross revenue information, let's clean the two datasets.
- Check if there are null values. Be careful they may appear to be valid strings.
- Convert the columns to the appropriate formats. In particular handle:
    - Released
    - Runtime
    - year
    - imdbRating
    - imdbVotes
- Merge the data from the two datasets into a single one

## 5.d Text vectorization

There are several columns in the data that contain a comma separated list of items, for example the Genre column and the Actors column. Let's transform those to binary columns using the count vectorizer from scikit learn.

Append these columns to the merged dataframe.

**Hint:** In order to get the actors name right, you'll have to set the `token_pattern` parameter in `CountVectorizer` to u'(?u)\\w+\.?\\w?\.? \\w+'. Can you see why? How does this differ from the default?

## Bonus:

- What are the top 10 grossing movies?
- Who are the 10 actors that appear in the most movies?
- What's the average grossing of the movies in which each of these actors appear?
- What genre is the oldest movie?
